File: app/src/main/java/cat/teknos/pokedex/ItemListFragment.java
package cat.teknos.pokedex;

import android.content.ClipData;
import android.content.ClipDescription;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import com.google.gson.GsonBuilder;

import cat.teknos.pokedex.api.Pokemon;
import cat.teknos.pokedex.api.PokemonFetchResults;
import cat.teknos.pokedex.databinding.FragmentItemListBinding;
import cat.teknos.pokedex.databinding.ItemListContentBinding;
import cat.teknos.pokedex.placeholder.PlaceholderContent;
import cat.teknos.pokedex.service.PokemonAPIService;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

import java.util.ArrayList;
import java.util.List;

public class ItemListFragment extends Fragment {

    private FragmentItemListBinding binding;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        binding = FragmentItemListBinding.inflate(inflater, container, false);
        return binding.getRoot();

    }

    @Override
    public void onViewCreated(@NonNull View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://pokeapi.co/api/v2/")
                .addConverterFactory(GsonConverterFactory.create(
                        new GsonBuilder().serializeNulls().create()
                ))
                .build();
        PokemonAPIService pokemonApiService = retrofit.create(PokemonAPIService.class);
        Call<PokemonFetchResults> call = pokemonApiService.getPokemons();
        call.enqueue(new Callback() {

            @Override
            public void onResponse(@NonNull Call call, @NonNull Response response) {
                if (response.isSuccessful()) {
                    assert response.body() != null;
                    ArrayList<Pokemon> pokemonList =((PokemonFetchResults) response.body()).getResults();
                } else {
                    Log.d("Error:", "Couldn't get list");
                }
            }

            @Override
            public void onFailure(@NonNull Call call, @NonNull Throwable t) {
                Log.d("Error:", t.toString());
            }
        });

        RecyclerView recyclerView = binding.itemList;
        View itemDetailFragmentContainer = view.findViewById(R.id.item_detail_nav_container);

        View.OnClickListener onClickListener = itemView -> {
            PlaceholderContent.PlaceholderItem item =
                    (PlaceholderContent.PlaceholderItem) itemView.getTag();
            Bundle arguments = new Bundle();
            arguments.putString(ItemDetailFragment.ARG_ITEM_ID, item.id);
            if (itemDetailFragmentContainer != null) {
                Navigation.findNavController(itemDetailFragmentContainer)
                        .navigate(R.id.fragment_item_detail, arguments);
            } else {
                Navigation.findNavController(itemView).navigate(R.id.show_item_detail, arguments);
            }
        };

        View.OnContextClickListener onContextClickListener = itemView -> {
            PlaceholderContent.PlaceholderItem item =
                    (PlaceholderContent.PlaceholderItem) itemView.getTag();
            Toast.makeText(
                    itemView.getContext(),
                    "Context click of item " + item.id,
                    Toast.LENGTH_LONG
            ).show();
            return true;
        };

        setupRecyclerView(recyclerView, onClickListener, onContextClickListener);
    }

    private void setupRecyclerView(
            RecyclerView recyclerView,
            View.OnClickListener onClickListener,
            View.OnContextClickListener onContextClickListener
    ) {

        recyclerView.setAdapter(new SimpleItemRecyclerViewAdapter(
                PlaceholderContent.ITEMS,
                onClickListener,
                onContextClickListener
        ));
    }

    public static class SimpleItemRecyclerViewAdapter
            extends RecyclerView.Adapter<SimpleItemRecyclerViewAdapter.ViewHolder> {

        private final List<PlaceholderContent.PlaceholderItem> mValues;
        private final View.OnClickListener mOnClickListener;
        private final View.OnContextClickListener mOnContextClickListener;

        SimpleItemRecyclerViewAdapter(List<PlaceholderContent.PlaceholderItem> items,
                                      View.OnClickListener onClickListener,
                                      View.OnContextClickListener onContextClickListener) {
            mValues = items;
            mOnClickListener = onClickListener;
            mOnContextClickListener = onContextClickListener;
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

            ItemListContentBinding binding = ItemListContentBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new ViewHolder(binding);

        }

        @Override
        public void onBindViewHolder(final ViewHolder holder, int position) {
            holder.mIdView.setText(mValues.get(position).id);
            holder.mContentView.setText(mValues.get(position).content);

            holder.itemView.setTag(mValues.get(position));
            holder.itemView.setOnClickListener(mOnClickListener);
            holder.itemView.setOnContextClickListener(mOnContextClickListener);
            holder.itemView.setOnLongClickListener(v -> {
                ClipData.Item clipItem = new ClipData.Item(mValues.get(position).id);
                ClipData dragData = new ClipData(
                        ((PlaceholderContent.PlaceholderItem) v.getTag()).content,
                        new String[]{ClipDescription.MIMETYPE_TEXT_PLAIN},
                        clipItem
                );

                v.startDragAndDrop(
                        dragData,
                        new View.DragShadowBuilder(v),
                        null,
                        0
                );
                return true;
            });
        }

        @Override
        public int getItemCount() {
            return mValues.size();
        }

        static class ViewHolder extends RecyclerView.ViewHolder {
            final TextView mIdView;
            final TextView mContentView;

            ViewHolder(ItemListContentBinding binding) {
                super(binding.getRoot());
                mIdView = binding.idText;
                mContentView = binding.content;
            }

        }
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/cat/teknos/pokedex/ItemListFragment.kt
package cat.teknos.pokedex

import android.content.ClipData
import android.content.ClipDescription
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.View.DragShadowBuilder
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.navigation.Navigation.findNavController
import androidx.recyclerview.widget.RecyclerView
import cat.teknos.pokedex.api.PokemonFetchResults
import cat.teknos.pokedex.databinding.FragmentItemListBinding
import cat.teknos.pokedex.databinding.ItemListContentBinding
import cat.teknos.pokedex.placeholder.PlaceholderContent
import cat.teknos.pokedex.placeholder.PlaceholderContent.PlaceholderItem
import cat.teknos.pokedex.service.PokemonAPIService
import com.google.gson.GsonBuilder
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class ItemListFragment : Fragment() {

    private var binding: FragmentItemListBinding? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentItemListBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val retrofit = Retrofit.Builder()
            .baseUrl("https://pokeapi.co/api/v2/")
            .addConverterFactory(
                GsonConverterFactory.create(
                    GsonBuilder().serializeNulls().create()
                )
            )
            .build()

        val pokemonApiService = retrofit.create(
            PokemonAPIService::class.java
        )

        val call = pokemonApiService.pokemons
        call?.enqueue(object : Callback<PokemonFetchResults?> {
            override fun onResponse(
                call: Call<PokemonFetchResults?>,
                response: Response<PokemonFetchResults?>
            ) {
                if (response.isSuccessful) {
                    assert(response.body() != null)
                    response.body()!!.results
                } else {
                    Log.d("Error:", "Couldn't get list")
                }
            }

            override fun onFailure(call: Call<PokemonFetchResults?>, t: Throwable) {
                TODO("Not yet implemented")
            }
        })

        val recyclerView = binding!!.itemList
        val itemDetailFragmentContainer = view.findViewById<View>(R.id.item_detail_nav_container)

        val onClickListener = View.OnClickListener { itemView: View ->
            val item = itemView.tag as PlaceholderItem
            val arguments = Bundle()
            arguments.putString(ItemDetailFragment.ARG_ITEM_ID, item.id)
            if (itemDetailFragmentContainer != null) {
                findNavController(itemDetailFragmentContainer)
                    .navigate(R.id.fragment_item_detail, arguments)
            } else {
                findNavController(itemView).navigate(R.id.show_item_detail, arguments)
            }
        }

        val onContextClickListener = View.OnContextClickListener { itemView: View ->
            val item = itemView.tag as PlaceholderItem
            Toast.makeText(
                itemView.context,
                "Context click of item " + item.id,
                Toast.LENGTH_SHORT
            ).show()
            true
        }
        setupRecyclerView(recyclerView, onClickListener, onContextClickListener)
    }

    private fun setupRecyclerView(
        recyclerView: RecyclerView,
        onClickListener: View.OnClickListener,
        onContextClickListener: View.OnContextClickListener
    ) {
        recyclerView.adapter = SimpleItemRecyclerViewAdapter(
            PlaceholderContent.ITEMS,
            onClickListener,
            onContextClickListener
        )
    }

    class SimpleItemRecyclerViewAdapter internal constructor(
        private val mValues: List<PlaceholderItem>,
        private val mOnClickListener: View.OnClickListener,
        private val mOnContextClickListener: View.OnContextClickListener
    ) : RecyclerView.Adapter<SimpleItemRecyclerViewAdapter.ViewHolder>() {
        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val binding =
                ItemListContentBinding.inflate(LayoutInflater.from(parent.context), parent, false)
            return ViewHolder(binding)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            holder.mIdView.text = mValues[position].id
            holder.mContentView.text = mValues[position].content
            holder.itemView.tag = mValues[position]
            holder.itemView.setOnClickListener(mOnClickListener)
            holder.itemView.setOnContextClickListener(mOnContextClickListener)
            holder.itemView.setOnLongClickListener { v: View ->
                val clipItem = ClipData.Item(mValues[position].id)
                val dragData = ClipData(
                    (v.tag as PlaceholderItem).content,
                    arrayOf(ClipDescription.MIMETYPE_TEXT_PLAIN),
                    clipItem
                )
                v.startDragAndDrop(
                    dragData,
                    DragShadowBuilder(v),
                    null,
                    0
                )
                true
            }
        }

        override fun getItemCount(): Int {
            return mValues.size
        }

        class ViewHolder(binding: ItemListContentBinding) : RecyclerView.ViewHolder(binding.root) {
            val mIdView: TextView
            val mContentView: TextView

            init {
                mIdView = binding.idText
                mContentView = binding.content
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}